### ISSUE TITLE
Regarding mean and other reductions of SparseTensor

### ISSUE THREAD

**Issue by flandolfi** [rusty1s/pytorch_sparse#159]
Hi Matthias,

I was working with `SparseTensor`s and it just occurred to me that `mean()` does not actually compute the mean, specifically:
```python
In [1]: from torch_sparse import SparseTensor

In [2]: SparseTensor.eye(10).mean(-1)
Out[2]: tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [3]: import torch

In [4]: torch.eye(10).mean(-1)
Out[4]: tensor([0.1000, 0.1000, 0.1000, 0.1000, 0.1000, 0.1000, 0.1000, 0.1000, 0.1000, 0.1000])
```
Now, this is not really an issue for me—in fact, `mean()` computes what I really needed—but rather a question: shouldn't `SparseTensor` work as a sparse replacement for PyTorch's `Tensor`?

I know this will probably break some code (mine for sure), but maybe `SparseTensor` should behave like a normal `Tensor` (similarly to `scipy.sparse` matrices), and users should be able to decide if use the classical `mean` reduction or `torch_scatter`'s `scatter_mean` on non-zero indices. 

Notice that this argument is also relevant to other reduction, e.g.,
```python
In [5]: SparseTensor.eye(10).min(-1)
Out[5]: tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [6]: torch.eye(10).min(-1)
Out[6]: 
torch.return_types.min(
values=tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
indices=tensor([1, 0, 0, 0, 0, 0, 0, 0, 0, 0]))
```
One way to keep both functions could be using `mean()`, `min()`, and `max()` as the `Tensor` reductions, while defining `nnz_mean()`, `nnz_min()` and `nnz_max()` respectively for the current sparse reductions.

Let me know what you think about it! :)

Best regards,
Francesco

**Comment by rusty1s** [comment]
Yeah, this is the classic inconsistency between sparse matrices and dense matrices. In fact, we define each value that is not included in the sparse matrix as "undefined" rather than "zero". For our use-case (adjacency matrices in Graph Neural Networks) this is the better approach, as we generally only want to compute "reductions" over the set of existing edges. We do not necessarily intend the `SparseTensor` class to be a drop-in replacement for the original `torch.Tensor`.

**Comment by flandolfi** [comment]
Understandable, thanks for the clarification! I'll close this "issue".
